src/forwarder.rs

```diff
@@ -3,121 +3,243 @@ use std::{
+                        // handle shutdown (avoid using sleep since it will hang under SIGINT)
+                        recv(shutdown_receiver) -> _ => {
+                            break;
+                        }
+                    }
+                }
+                info!("Exiting forwarder thread {thread_id}.");
+            })
+            .unwrap();
+
+        [listen_thread, send_thread]
+    })
+    .collect::<Vec<JoinHandle<()>>>()
+}
+
 /// broadcasts same packet to multiple recipients
 /// for best performance, connect sockets to their destinations
 /// Returns Err when unable to receive packets.
-fn send_multiple_destination_from_receiver(
-    receiver: &PacketBatchReceiver,
-    outgoing_sockets: &Arc<Vec<(UdpSocket, SocketAddr)>>,
-    successful_shred_count: &Arc<AtomicU64>,
-    failed_shred_count: &Arc<AtomicU64>,
+fn recv_from_channel_and_send_multiple_dest(
+    maybe_packet_batch: Result<PacketBatch, RecvError>,
+    deduper: &Arc<RwLock<Deduper>>,
+    send_socket: &UdpSocket,
+    local_dest_sockets: &Arc<Vec<SocketAddr>>,
+    metrics: &Arc<ShredMetrics>,
 ) -> Result<(), ShredstreamProxyError> {
-    let packet_batch = match receiver.recv_timeout(Duration::from_secs(1)) {
+    let packet_batch = match maybe_packet_batch {
         Ok(x) => Ok(x),
-        Err(RecvTimeoutError::Timeout) => return Ok(()),
-        Err(e) => Err(ShredstreamProxyError::StreamerError(StreamerError::from(e))),
+        Err(e) => Err(ShredstreamProxyError::RecvError(e)),
     }?;
+    metrics
+        .agg_received
+        .fetch_add(packet_batch.len() as u64, Ordering::Relaxed);
+    debug!(
+        "Got batch of {} packets, total size in bytes: {}",
+        packet_batch.len(),
+        packet_batch.iter().map(|x| x.meta.size).sum::<usize>()
+    );
+    let mut packet_batch_vec = vec![packet_batch];
 
-    outgoing_sockets
-        .iter()
-        .for_each(|(outgoing_socket, outgoing_socketaddr)| {
-            let packets = packet_batch
-                .iter()
-                .filter(|pkt| !pkt.meta.discard())
-                .filter_map(|pkt| {
-                    let data = pkt.data(..)?;
-                    let addr = outgoing_socketaddr;
-                    Some((data, addr))
-                })
-                .collect::<Vec<_>>();
-
-            match batch_send(outgoing_socket, &packets) {
-                Ok(_) => {
-                    successful_shred_count.fetch_add(packets.len() as u64, Ordering::SeqCst);
-                }
-                Err(SendPktsError::IoError(err, num_failed)) => {
-                    failed_shred_count.fetch_add(num_failed as u64, Ordering::SeqCst);
-                    error!(
-                        "Failed to send batch of size {} to {outgoing_socket:?}. {num_failed} packets failed. Error: {err}",
-                        packets.len()
-                    );
-                }
+    let num_deduped = deduper.read().unwrap().dedup_packets_and_count_discards(
```

Comment:
how much time is it spending waiting for lock here?

almost wonder if we should do:
[multiple sockets] -> [single dedup thread] -> multiple senders

Reply:
i think it should be minimal, since only once every 2 secs we acquire a write lock to reset the stats. otherwise everything uses a read lock

Reply:
write: https://github.com/jito-labs/shredstream-proxy/blob/8eac25fdd15a13f1c0b3d6989fd142c14a40773b/src/forwarder.rs#L282

Reply:
oh nice, wfm